m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -556,26 +556,31 @@ public void run() {
                 }
             }
 
-            state.set(ProcessorState.SHUTDOWN);
-
             LOG.warn("Processor shutting down");
 
             shutdown();
+            state.set(ProcessorState.SHUTDOWN);
 
             LOG.warn("Processor shut down");
         }
 
         private void shutdown() {
-            // Drain queue of any remaining metrics submitted prior to shutdown;
-            runNoThrow(this::drainQueue);
-            // Flush remaining buffers at last (best effort)
-            runNoThrow(this::flushBuffered);
+            if (isShutdown.get()) {
+                // If the processor hit an exception, another processor will be created to resume work on the queue,
+                // so leave the queue as is.
+                // Drain queue of any remaining metrics submitted prior to shutdown;
+                runNoThrow(this::drainQueue);
+                // Flush remaining buffers at last (best effort)
+                runNoThrow(this::flushBuffered);
+            }
 
             // Close transport
             transport.close();
 
-            // Count down shutdown latch to notify reporter
-            processorsShutdownLatch.countDown();
+            if (isShutdown.get()) {
```

Comment:
Instead, please replace the latch w/ `Semaphore`

Reply:
Sure.